### PR TITLE
Add group deleted applier

### LINKED ISSUE #23485: Apply `Group.DELETED` event
<h2>Description</h2>

In Zeebe we modify our state by applying events. We need to create a new `GroupDeletedApplier`:

1. Create the class in the same package as our other appliers
2. Implement `TypedEventApplier<GroupIntent, GroupRecord>`
3. Iterate over all entities of the `ENTITIES_BY_GROUP` ColumnFamily that have this `groupKey`
    - For each entity that is a USER:
        - Get the user by its key from `UserState`
        - Remove the Group from the `PersistedUser`
        - Update the PersistedUser in the state
     - For each entity that is a MAPPING (**Ignore if mappings aren't implemented yet**):
         - Get the mapping by its key from the `MappingState`
         - Remove the Group from the `PersistedMapping`
         - Update the `PeristedMapping` in the state
4. Iterate over all Authorizations in the `AuthorizationState` where the `ownerKey` is this `groupKey`
    - Delete any that are found 
5. Remove this Group from the `OWNER_TYPE_BY_OWNER_KEY` ColumnFamily in the `AuthorizationState`
6. Add a method to to the `MutableGroupState` interface to delete a group and implement this method in the `DbGroupState`:
    - Wrap the `groupKey`
    - Wrap the `groupName`
    - Delete from `GROUP_BY_NAME` ColumnFamily
    - Delete all entires for this group from the `ENTITY_BY_GROUP` ColumnFamily
    - Delete group from the `GROUPS` ColumnFamily
    - Unit test this method
7. In the `GroupUpdatedApplier` call this update method of the `DbGroupState`
8. Unit test all the new state methods

ℹ️ Note that we don't send individual events to remove entities from the group when we delete it. While this may be nice, this can quickly get out of hand if there's a lot of entities assigned to this group. There could be thousands of follow-up events which significantly increases the likelihood of running into `ExceededBatchRecordSizeException`

### PR DESCRIPTION
**Pull request by koevskinikola**
## Description

Provide an applier for `GroupIntent.DELETED` events, and the appropriate `GroupState` CRUD methods that are necessary for the data to be applied to the `DbGroupState`.

## Checklist

- [ ] for CI changes:
  - [ ] structural/foundational changes signed off by [CI DRI](https://github.com/cmur2)
  - [ ] [ci.yml](https://github.com/camunda/camunda/blob/main/.github/workflows/ci.yml) modifications comply with ["Unified CI" requirements](https://github.com/camunda/camunda/wiki/CI-&-Automation#workflow-inclusion-criteria)

## Related issues

closes #23485

### CODE REVIEW

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/appliers/GroupDeletedApplier.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+        case EntityType.MAPPING:
+          removeGroupFromState(key, entry.getValue(), mappingState::removeGroup);
+          break;
+        default:
+          throw new IllegalStateException(
+              String.format(
+                  "Expected to remove entity '%d' for group '%d', but entities of type '%s' are not supported by groups.",
+                  value.getEntityKey(), key, value.getEntityType()));
+      }
+    }
```
Comment (nicpuppa) [inline zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/appliers/GroupDeletedApplier.java:56]: 🔧 The for can be removed and you can use an enhanced switch

```
entitiesByType.forEach((entityType, entityIds) -> {
        BiConsumer<Long, Long> removalFunction = getRemovalFunction(entityType, groupKey, groupRecord);
        entityIds.forEach(entityId -> removalFunction.accept(entityId, groupKey));
    });
    
 private BiConsumer<Long, Long> getRemovalFunction(EntityType entityType, long groupKey, GroupRecord groupRecord) {
    return switch (entityType) {
        case EntityType.USER -> userState::removeGroup;
        case EntityType.MAPPING -> mappingState::removeGroup;
        default ->     throw new IllegalStateException(String.format("Entity type '%s' is unsupported for group '%d'. Expected entity '%d' not found in group.", entityType, groupKey, groupRecord.getEntityKey()
    ));
    };
}
```

File: zeebe/engine/src/test/java/io/camunda/zeebe/engine/state/group/GroupStateTest.java [hunk 2]
```diff
@@ -179,4 +179,30 @@ void shouldRemoveEntity() {
...
+    groupState.addEntity(groupKey, groupRecord);
+    groupRecord.setEntityKey(3L).setEntityType(EntityType.MAPPING);
+    groupState.addEntity(groupKey, groupRecord);
+
+    // when
+    groupState.delete(groupKey);
+
+    // then
+    final var group = groupState.get(groupKey);
+    assertThat(group.isPresent()).isFalse();
```
Comment (nicpuppa) [inline zeebe/engine/src/test/java/io/camunda/zeebe/engine/state/group/GroupStateTest.java:200]:
```suggestion
    assertThat(group).isEmpty();
```

File: zeebe/engine/src/test/java/io/camunda/zeebe/engine/state/group/GroupStateTest.java [hunk 3]
```diff
@@ -179,4 +179,30 @@ void shouldRemoveEntity() {
...
+
+    final var groupKeyByName = groupState.getGroupKeyByName(groupName);
+    assertThat(groupKeyByName.isPresent()).isFalse();
```
Comment (nicpuppa) [inline zeebe/engine/src/test/java/io/camunda/zeebe/engine/state/group/GroupStateTest.java:203]:
```suggestion
    assertThat(groupKeyByName).isEmpty();
```

**Review by nicpuppa — APPROVED** [review]
Please have a look at the suggestions 🙏 

@koevskinikola if this is the last applier for Group, please remove this [line](https://github.com/camunda/camunda/blob/f4cdbdf9382db04d89502a9500803c5ace953589/zeebe/engine/src/test/java/io/camunda/zeebe/engine/state/appliers/EventAppliersTest.java#L175)

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/appliers/GroupDeletedApplier.java (hunk 1 above)
Comment (koevskinikola) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/appliers/GroupDeletedApplier.java:56]: And 5 mins ago I added a similar enhance switch when Implementing the `GroupAddEntityProcessor` 😄

**Pull request merged**